### PR TITLE
Support Lexicographic optimization in Blendsearch

### PR DESCRIPTION
**Pull request by skzhang1**
## Why are these changes needed?

We plan to support lexicographic optimization in BlendSearch. This pull request is primarily intended to track the implementation progress. We encourage everyone to provide feedback in this pull request regarding both implementation suggestions and application scenarios.

**TODO:**
[ ] Implement a easier ECI calculation method
[ ] Force the global search algorithm to random search in lexicographic optimization.

## Related issue number

N/A

## Checks

- [ ] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR, or I've made sure [lint with flake8](https://github.com/microsoft/FLAML/blob/816a82a1155b4de4705b21a615ccdff67c6da379/.github/workflows/python-package.yml#L54-L59) output is two 0s.
- [ ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by skzhang1** [comment]
> Is it better to create a new subclass than modifying the original BlendSearch? It's an open question.

I think we do not need to do it for now. Because LexiFlow is also not independent of flow2 in the current version.

File: flaml/tune/searcher/flow2.py [hunk 1]
```diff
@@ -312,6 +314,24 @@ def denormalize(self, config):
         """denormalize each dimension in config from [0,1]."""
         return denormalize(config, self._space, self.best_config, self.incumbent, self._random)
 
+    def _get_lexico_bound(self, metric, mode):
```
Comment (sonichi) [inline flaml/tune/searcher/flow2.py:317]: The same function appears in multiple places, which is not a good design.

File: flaml/tune/searcher/search_thread.py [hunk 2]
```diff
@@ -76,26 +92,88 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
...
+        if not self.lexico_objectives:
+            if self._better(self.obj_best1, self.obj_best2):
+                self.speed = (
+                    (self.obj_best2 - self.obj_best1)
+                    / self.running
+                    / (max(self.cost_total - self.cost_best2, self._eps))
+                )
+            else:
+                self.speed = 0
+        elif self._search_alg._histories:
```
Comment (sonichi) [inline flaml/tune/searcher/search_thread.py:155]: don't access private fields

File: test/tune/test_lexiflow.py
Comment (sonichi) [inline test/tune/test_lexiflow.py:1]: Do you have a test for BlendSearch

**Review by sonichi — COMMENTED** [review]
This PR involves many big changes. There are many changes I don't understand. Could you summarize the changes first? Please also justify why a particular change should be done in that place. We do not want to add unnecessary changes or break the original functionality.
Later, I'll also request adding comments for all the non-obvious changes. Let's get a high-level understanding first.

File: flaml/tune/searcher/blendsearch.py [hunk 3]
```diff
@@ -194,9 +202,15 @@ def __init__(
         self._config_constraints = config_constraints
         self._metric_constraints = metric_constraints
         if metric_constraints:
-            assert all(x[1] in ["<=", ">="] for x in metric_constraints), "sign of metric constraints must be <= or >=."
-            # metric modified by lagrange
-            metric += self.lagrange
+            if self.lexico_objectives:
+                self._metric_constraints = None
+                logger.info("Do not support providing metric_constraints in lexicographic optimization for now.")
```
[the remark this answers is not in the document]
Comment (skzhang1) [reply flaml/tune/searcher/blendsearch.py]: resolved

File: flaml/tune/searcher/suggestion.py [hunk 4]
```diff
@@ -739,3 +921,100 @@ def resolve_value(domain: Domain) -> ot.distributions.BaseDistribution:
         values = {"/".join(path): resolve_value(domain) for path, domain in domain_vars}
 
         return values
+
+
+class LexiGlobalSearch(OptunaSearch):
```
[the remark this answers is not in the document]
Comment (skzhang1) [reply flaml/tune/searcher/suggestion.py]: resolved

File: flaml/tune/searcher/blendsearch.py [hunk 5]
```diff
@@ -643,11 +760,50 @@ def _expand_admissible_region(self, lower, upper, space):
...
+                elif obj_1[k_metr] < obj_2[k_metr]:
+                    return False
+                else:
+                    return True
+        else:
+            if obj_1 > obj_2:
+                return True
+            else:
+                return False
+
```
[the remark this answers is not in the document]
Comment (skzhang1) [reply flaml/tune/searcher/blendsearch.py]: resolved

File: flaml/tune/searcher/blendsearch.py [hunk 6]
```diff
@@ -643,11 +760,50 @@ def _expand_admissible_region(self, lower, upper, space):
...
+        else:
+            result[self._ls.metric] = result[self._ls.metric] * self._ls.metric_op
+        return result
+
+    def _lexico_sort(self, arr: list):
+        n = len(arr)
+        for i in range(n):
+            for j in range(0, n - i - 1):
+                if self._lexico_(arr[j], arr[j + 1]):
+                    arr[j], arr[j + 1] = arr[j + 1], arr[j]
```
[the remark this answers is not in the document]
Comment (skzhang1) [reply flaml/tune/searcher/blendsearch.py]: resolved

File: flaml/tune/searcher/blendsearch.py [hunk 7]
```diff
@@ -664,6 +820,12 @@ def on_trial_result(self, trial_id: str, result: Dict):
             return
         if result and self._metric_constraints:
             result[self._metric + self.lagrange] = result[self._metric]
+            if self.lexico_objectives:
+                if self._histories is None:
+                    self._histories, self._f_best = defaultdict(list), {}
+                for k_metric, k_mode in zip(self.lexico_objectives["metrics"], self.lexico_objectives["modes"]):
+                    self._histories[k_metric].append(result[k_metric] if k_mode == "min" else -1 * result[k_metric])
+                self.update_fbest()
```
[the remark this answers is not in the document]
Comment (skzhang1) [reply flaml/tune/searcher/blendsearch.py]: I think metric constraints should be provided via a target in lexicographic_objectives

File: flaml/tune/searcher/blendsearch.py [hunk 8]
```diff
@@ -806,10 +972,18 @@ def _violate_config_constriants(self, config, config_signature):
...
+                        self._metric: np.inf * self._ls.metric_op,
+                        "time_total_s": 1,
+                    }
+                else:
+                    for k_metric, k_mode in zip(self.lexico_objectives["metrics"], self.lexico_objectives["modes"]):
+                        self._result[config_signature] = {}
+                        self._result[config_signature][k_metric] = {
+                            self._metric: np.inf * -1 if k_mode == "max" else np.inf
+                        }
+                        self._result[config_signature]["time_total_s"] = 1
```
[the remark this answers is not in the document]
Comment (skzhang1) [reply flaml/tune/searcher/blendsearch.py]: resolved

File: flaml/tune/searcher/search_thread.py (hunk 2 above)
Comment (skzhang1) [reply flaml/tune/searcher/search_thread.py]: resolved

File: test/tune/test_lexiflow.py
Comment (skzhang1) [reply test/tune/test_lexiflow.py:1]: resolved. I added a simple test function for now

File: flaml/tune/tune.py [hunk 9]
```diff
@@ -511,7 +492,11 @@ def easy_objective(config):
                 logger.warning("Using CFO for search. To use BlendSearch, run: pip install flaml[blendsearch]")
             metric = metric or DEFAULT_METRIC
         else:
-            SearchAlgorithm = CFO
+            assert "lexico_algorithm" in lexico_objectives and lexico_objectives["lexico_algorithm"] in [
+                "CFO",
+                "BlendSearch",
+            ], 'When performing lexicographic optimization, "lexico_algorithm" should be provided in lexicographic objectives (CFO or BlendSearch).'
+            SearchAlgorithm = CFO if lexico_objectives["lexico_algorithm"] == "CFO" else BlendSearch
```
[the remark this answers is not in the document]
Comment (skzhang1) [reply flaml/tune/tune.py:499]: I think it is simpler for users to specify algorithms from lexico_objectives. Because `search_alg`  in tune represents algorithm instance rather than strings. We have also not tried passing instances of LexiFlow to use tune before.

[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/tune/tune.py:499]: You can support str type `search_alg` in `tune` to make it easier. But don't create a new argument which is confusing.
`AutoML` uses `tune` by passing instance of CFO.

File: flaml/tune/searcher/blendsearch.py [hunk 10]
```diff
@@ -26,7 +28,7 @@
 from .flow2 import FLOW2
 from ..space import add_cost_to_space, indexof, normalize, define_by_run_func
 from ..result import TIME_TOTAL_S
-
+from ..utils import get_lexico_bound
```
Comment (qingyun-wu) [inline flaml/tune/searcher/blendsearch.py:31]: get_lexico_bound could be included in this file as a staticmethod of `BlendSearch`
